### PR TITLE
Changed ELKM to generate one meta data provider per layout algorithm. #9

### PR DESCRIPTION
**Pull request by le-cds**
With these changes, the ELKM editor generates one main layout meta data
provider that contains all declared options, as well as one layout meta
data provider for each registered layout algorithm. The latter always
contain IProperty constants for each supported layout option that can be
used to retrieve property values with the proper default values as
specified by the respective layout algorithm.

Signed-off-by: Christoph Daniel Schulze <cds@informatik.uni-kiel.de>

### CODE REVIEW

**Comment by le-cds** [comment]
You should probably limit yourself to reviewing the new grammar and the new JvmModelInferrer. I haven't removed the `duplicated` keyword yet, but will do so once you guys approve the merge request.

**Comment by uruuru** [comment]
For reviewing, it helps to split the actual changes and the re-generation of the code into two commits.

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend [hunk 1]
```diff
@@ -93,31 +99,28 @@ class MetaDataJvmModelInferrer extends AbstractModelInferrer {
             // 1. Public constants for all declared properties
             for (property : bundle.members.allPropertyDefinitions) {
                 val constant = property.toPropertyConstant
-                if (property.defaultValue !== null)
+                if (property.defaultValue !== null) {
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend]: We should either have a convention to always use braces (even in xtend files), or not change it in existing code.

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend [hunk 2]
```diff
@@ -293,35 +266,158 @@ class MetaDataJvmModelInferrer extends AbstractModelInferrer {
...
+        
+        // Create a class for the general property definitions to be used by clients as well as for the registration
+        // method that instantiates the ILayoutMetaDataProviders
+        acceptor.accept(algorithm.toClass(algorithm.qualifiedTargetClass)) [
+            // Implement the ILayoutMetaDataProvider interface and... BE DOCUMENTED!
+            superTypes += typeRef(ILayoutMetaDataProvider)
+            fileHeader = algorithm.bundle.documentation
+            documentation = algorithm.documentation
+            
+            // 1. Public constants for supported layout properties
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend]: we decided to call them 'options'

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend (hunk 1 above)
Comment (le-cds) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend]: I'm all for braces.

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend (hunk 2 above)
Comment (le-cds) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend]: Yup, I'll change that. I'll also take the opportunity to rename MdProperty to MdOption, as previously discussed.

**Pull request merged**